util/ranger/detacher.go

```diff
@@ -355,6 +355,112 @@ type DetachRangeResult struct {
+		for _, item := range cnfItems {
+			ok := detachColumnsFromItem(item, cols)
+			if !ok {
+				return false
+			}
+		}
+		return true
+	case ast.GE, ast.GT, ast.LT, ast.LE, ast.EQ, ast.NE:
+		if c, ok := sf.GetArgs()[0].(*expression.Column); ok {
+			if c.RetType.EvalType() == types.ETString && !collate.CompatibleCollate(c.RetType.Collate, collation) {
+				return false
+			}
+			if _, ok := sf.GetArgs()[1].(*expression.Constant); ok {
+				return true
+			}
+		}
+		if c, ok := sf.GetArgs()[1].(*expression.Column); ok {
+			if c.RetType.EvalType() == types.ETString && !collate.CompatibleCollate(c.RetType.Collate, collation) {
+				return false
+			}
+			if _, ok := sf.GetArgs()[0].(*expression.Constant); ok {
+				return true
+			}
+		}
+	case ast.In:
+		c, ok := sf.GetArgs()[0].(*expression.Column)
+		if !ok {
+			return false
+		}
+		if c.RetType.EvalType() == types.ETString && !collate.CompatibleCollate(c.RetType.Collate, collation) {
+			return false
+		}
+		for _, arg := range sf.GetArgs()[1:] {
+			if _, ok := arg.(*expression.Constant); !ok {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+// getDNFItems checks if sf is a logicOr function and all items in sf only contain logicAnd, In, LT, LE, GT, GE, EQ and NE.
+// If so, it will return item list. Otherwise, return nil.
+func getDNFItems(sf *expression.ScalarFunction, cols []*expression.Column) []expression.Expression {
+	if sf.FuncName.L != ast.LogicOr {
+		return nil
+	}
+	dnfItems := expression.FlattenDNFConditions(sf)
+	for _, item := range dnfItems {
+		ok := detachColumnsFromItem(item, cols)
+		if !ok {
+			return nil
+		}
+	}
+	return dnfItems
+}
+
+// expandDNF will expand one DNF in conditions if this DNF satisfies:
+// 		1. Only contains logicAnd, In, LT, LE, GT, GE, EQ and NE.
+//		2. For In function, it satisfies 'column in (constant list)'.
+// 		3. For LT, LE, GT, GE, EQ and NE functions, they satisfy one side is constant and another is column.
+// If the expansion is successful, it will return other conditions and all items expanded. Otherwise, return nil, nil.
+//
+// This is to correctly select the composite index in similar example
+//    'select * from t where a = 1 and ((b = 1 and c = 1) or (b = 2 and c = 2))'. (issue/14795)
+func expandDNF(conditions []expression.Expression, cols []*expression.Column) ([]expression.Expression, []expression.Expression) {
+	var resCond []expression.Expression
+	var dnfItems []expression.Expression
+	if conditions == nil {
+		return nil, nil
+	}
+	for _, cond := range conditions {
+		eqOrIn := getEqOrInColOffset(cond, cols)
+		if eqOrIn != -1 {
+			resCond = append(resCond, cond)
+		} else {
+			sf, ok := cond.(*expression.ScalarFunction)
+			if !ok {
+				return nil, nil
+			}
+			if sf.FuncName.L != ast.LogicOr || dnfItems != nil {
+				resCond = append(resCond, cond)
+				continue
+			}
+			dnfItems = getDNFItems(sf, cols)
+			if dnfItems == nil {
+				resCond = append(resCond, cond)
+			}
+		}
```

Comment:
Calling `getEqOrInColOffset` here is unnecessary, since if the function is `Or`, its `eqOrIn` must be -1.

Reply:
`getEqOrInColOffset` is unnecessary, because it repeats with `getDNFItems`. But "since if the function is Or, its eqOrIn must be -1" is incorrect.